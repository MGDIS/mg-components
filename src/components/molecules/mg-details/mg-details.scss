@import '../../../styles/a11y.scss';

.mg-details summary {
  display: flex;
  align-items: baseline;
  cursor: pointer;
}

.mg-details__toggle {
  margin-left: 1.5rem;
  flex-shrink: 0;
  display: flex;
  align-items: center;
  align-self: flex-start;
  gap: 0.6rem;
  min-height: calc(var(--font-size) * var(--line-height));
  mg-icon {
    align-self: self-end;
  }
}

.mg-details__details {
  margin-top: 1.5rem;
}
