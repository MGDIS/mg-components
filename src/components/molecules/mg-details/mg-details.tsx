import { Component, Event, EventEmitter, h, Prop, Watch } from '@stencil/core';

@Component({
  tag: 'mg-details',
  styleUrl: 'mg-details.scss',
  shadow: true,
})
export class MgDetails {
  /************
   * Internal *
   ************/

  // HTML selector
  private details: HTMLDetailsElement;

  /**************
   * Decorators *
   **************/

  /**
   * Displayed title when details are closed
   */
  @Prop() toggleClosed!: string;

  /**
   * Displayed title when details are opened
   */
  @Prop() toggleOpened!: string;
  @Watch('toggleClosed')
  @Watch('toggleOpened')
  validateTitles(newValue: string): void {
    if (newValue === undefined || newValue.trim() === '') {
      throw new Error('<mg-details> prop "toggleClosed" and "toggleOpened" must be defined.');
    }
  }

  /**
   * Hide summary element
   */
  @Prop() hideSummary = false;

  /**
   * Define if details are diplayed
   */
  @Prop({ mutable: true }) expanded = false;
  @Watch('expanded')
  handleExpanded(newValue: boolean): void {
    this.expandedChange.emit(newValue);
  }

  /**
   * Emmited event when expanded change
   */
  @Event({ eventName: 'expanded-change' }) expandedChange: EventEmitter<boolean>;

  /**
   * Handle details toggle
   *
   * @returns {void}
   */
  private handleToggle = (): void => {
    this.expanded = this.details.open;
  };

  /*************
   * Lifecycle *
   *************/

  /**
   * Check if component props are well configured on init
   *
   * @returns {void}
   */
  componentWillLoad(): void {
    this.validateTitles(this.toggleClosed);
    this.validateTitles(this.toggleOpened);
  }

  /**
   * Render
   *
   * @returns {HTMLElement} HTML Element
   */
  render(): HTMLElement {
    return (
      <details class="mg-details" onToggle={this.handleToggle} open={this.expanded} ref={el => (this.details = el as HTMLDetailsElement)}>
        <summary>
          <slot name="summary"></slot>
          <span class="mg-details__toggle">
            <mg-icon icon={this.expanded ? 'chevron-up' : 'chevron-down'}></mg-icon>
            <span class={{ 'sr-only': this.hideSummary }}>{this.expanded ? this.toggleOpened : this.toggleClosed}</span>
          </span>
        </summary>
        <div class="mg-details__details">
          <slot name="details"></slot>
        </div>
      </details>
    );
  }
}
