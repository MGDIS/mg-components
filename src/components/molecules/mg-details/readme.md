## Properties
 
![](./mg-details/doc/img/mg-details-components.png)

- **summary** and **details** are slots which allow html content
- **toggle** is defined by mg-icon plus a custom text
  - mg-icon _chevron-down_ is used for the close state
  - mg-icon _chevron-up_ is used for the open state

## Behavior

### Action

The complete zone including the _summary_ and the _toggle_ is clickable and toggles the state of the component.
 
![](./mg-details/doc/img/mg-details-behavior.png)

### Position

_Toggle_ follows the _summary_ and is always on the top right.
 
![](./mg-details/doc/img/mg-details-position.png)

### Responsive

In mobile resolution, the text of the _toggle_ is hidden
 
![](./mg-details/doc/img/mg-details-responsive.png)

## Specs

### Spacing
 
![](./mg-details/doc/img/mg-details-spacing.png)

<!-- Auto Generated Below -->


## Properties

| Property                    | Attribute       | Description                             | Type      | Default     |
| --------------------------- | --------------- | --------------------------------------- | --------- | ----------- |
| `expanded`                  | `expanded`      | Define if details are diplayed          | `boolean` | `false`     |
| `hideSummary`               | `hide-summary`  | Hide summary element                    | `boolean` | `false`     |
| `toggleClosed` _(required)_ | `toggle-closed` | Displayed title when details are closed | `string`  | `undefined` |
| `toggleOpened` _(required)_ | `toggle-opened` | Displayed title when details are opened | `string`  | `undefined` |


## Events

| Event             | Description                        | Type                   |
| ----------------- | ---------------------------------- | ---------------------- |
| `expanded-change` | Emmited event when expanded change | `CustomEvent<boolean>` |


## Dependencies

### Depends on

- [mg-icon](../../atoms/mg-icon)

### Graph
```mermaid
graph TD;
  mg-details --> mg-icon
  style mg-details fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
