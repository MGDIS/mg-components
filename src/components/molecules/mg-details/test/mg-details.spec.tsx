import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgDetails } from '../mg-details';

const getPage = args =>
  newSpecPage({
    components: [MgDetails],
    template: () => (
      <mg-details {...args}>
        <span slot="summary">Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.</span>
        <p slot="details">
          Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
          cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
        </p>
      </mg-details>
    ),
  });

describe('mg-details', () => {
  test.each([
    { toggleClosed: 'Show details', toggleOpened: 'Hide details' },
    { toggleClosed: 'Show details', toggleOpened: 'Hide details', expanded: true },
  ])('with args %s', async args => {
    const { root } = await getPage(args);
    expect(root).toMatchSnapshot();
  });

  test.each([{}, { toggleClosed: 'Show details' }, { toggleOpened: 'Hide details' }])('Should throw error when missing toogle messages : %s', async args => {
    expect.assertions(1);
    try {
      await getPage(args);
    } catch (err) {
      expect(err.message).toMatch('<mg-details> prop "toggleClosed" and "toggleOpened" must be defined.');
    }
  });

  test('Should toggle details on click', async () => {
    const page = await getPage({ toggleClosed: 'Show details', toggleOpened: 'Hide details' });

    const element = page.doc.querySelector('mg-details');
    const summary = element.shadowRoot.querySelector('summary');

    jest.spyOn(page.rootInstance.expandedChange, 'emit');

    summary.dispatchEvent(new CustomEvent('toggle', { bubbles: true }));
    await page.waitForChanges();

    expect(page.rootInstance.expandedChange.emit).toHaveBeenCalledTimes(1);
  });
});
