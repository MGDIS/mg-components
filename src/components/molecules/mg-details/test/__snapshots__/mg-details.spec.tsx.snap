// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-details with args {
  toggleClosed: 'Show details',
  toggleOpened: 'Hide details',
  expanded: true
} 1`] = `
<mg-details>
  <mock:shadow-root>
    <details class="mg-details" open="">
      <summary>
        <slot name="summary"></slot>
        <span class="mg-details__toggle">
          <mg-icon icon="chevron-up"></mg-icon>
          <span>
            Hide details
          </span>
        </span>
      </summary>
      <div class="mg-details__details">
        <slot name="details"></slot>
      </div>
    </details>
  </mock:shadow-root>
  <span slot="summary">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
  </span>
  <p slot="details">
    Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
</mg-details>
`;

exports[`mg-details with args {
  toggleClosed: 'Show details',
  toggleOpened: 'Hide details',
  hideSummary: true
} 1`] = `
<mg-details>
  <mock:shadow-root>
    <details class="mg-details">
      <summary>
        <slot name="summary"></slot>
        <span class="mg-details__toggle">
          <mg-icon icon="chevron-down"></mg-icon>
          <span class="sr-only">
            Show details
          </span>
        </span>
      </summary>
      <div class="mg-details__details">
        <slot name="details"></slot>
      </div>
    </details>
  </mock:shadow-root>
  <span slot="summary">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
  </span>
  <p slot="details">
    Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
</mg-details>
`;

exports[`mg-details with args { toggleClosed: 'Show details', toggleOpened: 'Hide details' } 1`] = `
<mg-details>
  <mock:shadow-root>
    <details class="mg-details">
      <summary>
        <slot name="summary"></slot>
        <span class="mg-details__toggle">
          <mg-icon icon="chevron-down"></mg-icon>
          <span>
            Show details
          </span>
        </span>
      </summary>
      <div class="mg-details__details">
        <slot name="details"></slot>
      </div>
    </details>
  </mock:shadow-root>
  <span slot="summary">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
  </span>
  <p slot="details">
    Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
</mg-details>
`;
